Implement initial framework to be able to start developing tests and …

qchaupds opened the pull request:
Implement initial framework to be able to start developing tests and improve error handling for data collection in a sub-directory

Closes NASA-PDS/validate#282 NASA-PDS/validate#283 NASA-PDS/validate#240 and NASA-PDS/validate#290


There is an existing integration test that gets run automatically.
This new capability allows tester/developer to add new tests for validation with the new cucumber framework
without having to change Java code.

Tested in DEV:

To test NASA-PDS/validate#282:

The test code for cucumber behavior testing will need to be compiled with:

% mvn package -DskipTests

The -DskipTests is to skip the (normal) existing integration tests.

A file will be created target/validate-1.25.0-SNAPSHOT-bin.zip by the command above and need to be unzipped with:

% cd target
% unzip validate-1.25.0-SNAPSHOT-bin.zip

A directory of .jar files will be created in validate-1.25.0-SNAPSHOT/lib

To run the cucumber testing assuming the version value in pom.xml for artifactId 'validate' is '1.25.0-SNAPSHOT'

% cd ../
% java -cp "target/test-classes:target/validate-1.25.0-SNAPSHOT/lib/*" io.cucumber.core.cli.Main target/test-classes/features

12 Scenarios (12 passed)
48 Steps (48 passed)
0m23.997s

To test NASA-PDS/validate#240:

% validate -R pds4.bundle  -r report_github240_bundle_valid.json   -s json -t src/test/resources/github240/valid/bundle_kaguya_derived.xml

% validate -R pds4.bundle  -r report_github240_bundle_invalid.json -s json -t src/test/resources/github240/invalid/bundle_kaguya_derived.xml

Commit:
Implement initial framework to be able to start developing tests and improve error handling for data collection in a sub-directory

1. Add github240 test resources to src/test/resources
2. Add cucumber dependencies to pom.xml
3. Add cucumber feature artifacts to src/test/resources for initial cucumber behavior testing
4. Migrate subset of integration tests to cucumber tests (via the validate.feature file) to new class src/test/java/cucumber/StepDefs.java
5. Change report type to WARNING and message in PDS4Problems.java
6. Add debug to SubdirectoryNamingRule.java

Refs #282 https://github.com/NASA-PDS/validate/issues/282 Implement initial framework to be able to start developing tests
     #283 https://github.com/NASA-PDS/validate/issues/283 Implement initial test to exercise framework
     #240 https://github.com/NASA-PDS/validate/issues/240 RuleUnexpected error for data collection in a sub-directory
     #290 https://github.com/NASA-PDS/validate/issues/290 Migrate subset of existing regression tests to cucumber behavioral testing

jordanpadams reviewed (changes requested):
@qchaupds I think I would prefer if we ran these cucumber tests within a JUnit Runner class so they actually execute during the maven test phase. if we have any tests that are duplicated within the `ValidationIntegrationTests` let's just remove them.

@tloubrieu-jpl does that seem reasonable to you?

Commit:
Add logic to allow cucumber tests to run from maven.   Running from command line is also available.

1. Add profile 'cucumber-test-run' to allow maven to run cucumber test to pom.xml
2. Add new java class FileFinder.java to search for resource validation-commands.xml
3. Add logic to find validation-commands.xml due to resource null when ran from maven to LocationValidator.java
4. Add logic to find validation-commands.xml due to resource null when ran from maven to ValidationResourceManager.java

Refs #282 https://github.com/NASA-PDS/validate/issues/282 Implement initial framework to be able to start developing tests
     #283 https://github.com/NASA-PDS/validate/issues/283 Implement initial test to exercise framework
     #240 https://github.com/NASA-PDS/validate/issues/240 RuleUnexpected error for data collection in a sub-directory
     #290 https://github.com/NASA-PDS/validate/issues/290 Migrate subset of existing regression tests to cucumber behavioral testing

jordanpadams committed:
Cleanup pom and add junit-vintage-engine

jordanpadams committed:
Add cucumber test runner

jordanpadams reviewed (changes requested):
@qchaupds thanks for tryin to dig up a solution here. but we really needed this to work with `mvn test`. I was able to track down in the JUnit subsection under https://cucumber.io/docs/cucumber/api/?sbsearch=junit#running-cucumber, that Cucumber is based off of JUnit 4 and we are using JUnit 5. Because of this, we needed to add `junit-vintage-engine`. I then added a `CucumberTest` JUnit runner class. And it now appears to run nominally with `mvn test`. However, a few tests are failing. can you please take a look?

Commit:
Change report names to distinguish between integration test versus cucuumber test

jordanpadams committed:
Disable integration tests and rely on cucumber

per issues with validate caching, going to disable the integration tests for cucumber.

jordanpadams reviewed (changes requested):
@qchaupds so after poking around at this a bit, i can't figure out why the caching problem is still happening.

that being said, i don't think we should spend any more time debugging it. i've disabled ValidationIntegrationTests for now, and let's just migrate all of them to cucumber. once we fix the caching issue in validate next build, we can try to re-enable the tests to verify the fix.

thoughts?

Commit:
Migrate all tests from ValidationIntegrationTests.java to cucumber framework due to issue with running tests simultaneously.

1. Add test resource to github153 due to cucumber feature does not play well with spaces in file name.
2. Add more enumerated problem types and correct search for "validate#71" to build catalog file in StepDefs.java
3. Migrate all integration tests to validate.feature

Refs #282 https://github.com/NASA-PDS/validate/issues/282 Implement initial framework to be able to start developing tests
     #283 https://github.com/NASA-PDS/validate/issues/283 Implement initial test to exercise framework
     #240 https://github.com/NASA-PDS/validate/issues/240 RuleUnexpected error for data collection in a sub-directory
     #290 https://github.com/NASA-PDS/validate/issues/290 Migrate subset of existing regression tests to cucumber behavioral testing

Commit:
Add empty files to directories otherwise git will ignore them from repository.

1. Add empty files to 'data', 'browse' and 'calibration' dirctories.
2. Correct message from errors to warnings for "github 240 test 2" to validate.feature

Refs #282 https://github.com/NASA-PDS/validate/issues/282 Implement initial framework to be able to start developing tests
     #283 https://github.com/NASA-PDS/validate/issues/283 Implement initial test to exercise framework
     #240 https://github.com/NASA-PDS/validate/issues/240 RuleUnexpected error for data collection in a sub-directory
     #290 https://github.com/NASA-PDS/validate/issues/290 Migrate subset of existing regression tests to cucumber behavioral testing
